src/shell/components/FieldTypeNumber.tsx

```diff
@@ -0,0 +1,160 @@
+import { useEffect, useRef, useState } from "react";
+import { IconButton, Stack, TextField } from "@mui/material";
+import RemoveRoundedIcon from "@mui/icons-material/RemoveRounded";
+import AddRoundedIcon from "@mui/icons-material/AddRounded";
+
+type FieldTypeNumberProps = {
+  required: boolean;
+  name: string;
+  value: number;
+  onChange: (value: number, name: string) => void;
+  hasError: boolean;
+};
+export const FieldTypeNumber = ({
+  required,
+  value,
+  onChange,
+  name,
+  hasError,
+}: FieldTypeNumberProps) => {
+  const numberInputRef = useRef(null);
+  const [displayValue, setDisplayValue] = useState("0");
+  const [currentCursorPosition, setCurrentCursorPosition] = useState(0);
+  const [originalNumberOfCommas, setOriginalNumberOfCommas] = useState(0);
+
+  useEffect(() => {
+    setDisplayValue(addThousandsSeparator(value.toString()));
+  }, [value]);
+
+  useEffect(() => {
+    adjustCursorPosition(displayValue);
+  }, [displayValue]);
+
+  const modifyNumberValue = (action: "increment" | "decrement") => {
+    switch (action) {
+      case "increment":
+        onChange(value + 1, name);
+        break;
+
+      case "decrement":
+        onChange(value - 1, name);
+        break;
+
+      default:
+        break;
+    }
+  };
+
+  const adjustCursorPosition = (numStr: string) => {
```

Comment:
I’m actually no longer using the react-number format library here hence all the complexity and bugs since I have to handle all the cursor position etc manually. I’m actually considering using the library again since it takes care of all the formatting and resolves all of the bugs @zcolah has been raising for the past couple of days. Only downside to that library being that it automatically converts the exponents to numbers with leading 0s instead of showing the actual exponent.